cachito/workers/pkg_managers/pip.py

```diff
@@ -0,0 +1,321 @@
+
+        try:
+            version = get_top_level_attr(module_ast, attr_name)
+            log.debug("Found attribute %r in %r: %r", attr_name, module_name, version)
+            return version
+        except AttributeError as e:
+            log.debug("Could not find attribute in %r: %s", module_name, e)
+            return None
+
+    def _find_module(self, module_name, package_dir=None):
+        """
+        Try to find a module in the project directory and return path to source file.
+
+        :param str module_name: "import path" of module
+        :param dict[str, str] package_dir: same semantics as options.package_dir in setup.cfg
+
+        :rtype: Path or None
+        """
+        module_path = self._convert_to_path(module_name)
+        root_module = module_path.parts[0]
+
+        package_dir = package_dir or {}
+
+        if root_module in package_dir:
+            custom_path = Path(package_dir[root_module])
+            log.debug(f"Custom path set for root module %r: %r", root_module, str(custom_path))
+            # Custom path replaces the root module
+            module_path = custom_path.joinpath(*module_path.parts[1:])
+        elif "" in package_dir:
+            custom_path = Path(package_dir[""])
+            log.debug(f"Custom path set for all root modules: %r", str(custom_path))
+            # Custom path does not replace the root module
+            module_path = custom_path / module_path
+
+        full_module_path = self._ensure_local(module_path)
+
+        package_init = full_module_path / "__init__.py"
+        if package_init.is_file():
+            return package_init
+
+        module_py = Path(f"{full_module_path}.py")
+        if module_py.is_file():
+            return module_py
+
+        return None
+
+    def _convert_to_path(self, module_name):
+        """Check that module name is valid and covert to file path."""
+        parts = module_name.split(".")
+        if not parts[0]:
+            # Relative import (supported only to the extent that one leading '.' is ignored)
+            parts.pop(0)
+        if not all(self._name_re.fullmatch(part) for part in parts):
+            raise ValidationError(f"{module_name!r} is not an accepted module name")
+        return Path(*parts)
+
+    def _get_package_dirs(self):
+        """
+        Get options.package_dir and convert to dict if present.
+
+        https://github.com/pypa/setuptools/blob/ba209a15247b9578d565b7491f88dc1142ba29e4/setuptools/config.py#L264
+
+        :rtype: dict[str, str] or None
+        """
+        package_dir_value = self._get_option("options", "package_dir")
+        if package_dir_value is None:
+            return None
+
+        if "\n" in package_dir_value:
+            package_items = package_dir_value.splitlines()
+        else:
+            package_items = package_dir_value.split(",")
+
+        # Strip whitespace and discard empty values
+        package_items = filter(bool, (p.strip() for p in package_items))
```

Comment:
This is neat! It avoids the repetition seen in a pure list comprehension approach: 
```python
(p.strip() for p in package_items if p.strip())
```

Reply:
Right? That's pretty much exactly what setuptools does in their code, so of course I thought: "Ha! I can do you one better."